crux-mir/src/Mir/Language.hs

```diff
@@ -156,25 +163,51 @@ withMirLogging computation =
      in computation
 
 mainWithOutputConfig :: (Maybe Crux.OutputOptions -> OutputConfig MirLogging)
-                     -> BindExtraOverridesFn -> IO ExitCode
-mainWithOutputConfig mkOutCfg bindExtra =
+                     -> p -> BindExtraOverridesFn p -> IO ExitCode
+mainWithOutputConfig mkOutCfg customState bindExtra =
     withMirLogging $
     Crux.loadOptions mkOutCfg "crux-mir" Paths_crux_mir.version mirConfig
-        $ runTestsWithExtraOverrides bindExtra
+        $ runTestsWithExtraOverrides customState bindExtra
+
+type HasMirState p s = (MirPersonality p, MirState p ~ s)
+
+-- | Provides access to some custom state
+class MirPersonality p where
+  type MirState p
+  mirStateField :: Lens' p (MirState p)
+
+newtype CruxMir s sym = CruxMir { cruxMirState :: s }
+
+cruxMirNoState :: CruxMir () sym
+cruxMirNoState = CruxMir ()
+
+instance MirPersonality (CruxMir s sym) where
+  type MirState (CruxMir s sym) = s 
+  mirStateField = lens cruxMirState (const CruxMir)
+
+-- | This is used when we execute with concurrency
+instance MirPersonality p => MirPersonality (Explore.Exploration p alg ext ret sym) where
+  type MirState (Explore.Exploration p alg ext ret sym) = MirState p
+  mirStateField = Explore.personality . mirStateField
+
+mirState :: HasMirState (p sym) s => Lens' (C.SimState (p sym) sym ext r f a) s
```

Comment:
Why not generalize this to:

```suggestion
mirState :: HasMirState p s => Lens' (C.SimState p sym ext r f a) s
```

Reply:
I just pushed a changed that generalizes this in Crux

Reply:
Thanks! It's probably worth mentioning this change in the `crux` changelog. (Case in point: CI reveals that change broke some other code in `crux-llvm`.)